source/common/config/new_grpc_mux_impl.cc

```diff
@@ -0,0 +1,238 @@
+#include "common/config/new_grpc_mux_impl.h"
+
+#include "common/common/assert.h"
+#include "common/common/backoff_strategy.h"
+#include "common/common/token_bucket_impl.h"
+#include "common/config/utility.h"
+#include "common/protobuf/protobuf.h"
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Config {
+
+NewGrpcMuxImpl::NewGrpcMuxImpl(Grpc::RawAsyncClientPtr&& async_client,
+                               Event::Dispatcher& dispatcher,
+                               const Protobuf::MethodDescriptor& service_method,
+                               Runtime::RandomGenerator& random, Stats::Scope& scope,
+                               const RateLimitSettings& rate_limit_settings,
+                               const LocalInfo::LocalInfo& local_info)
+    : dispatcher_(dispatcher), local_info_(local_info),
+      grpc_stream_(this, std::move(async_client), service_method, random, dispatcher, scope,
+                   rate_limit_settings) {}
+
+Watch* NewGrpcMuxImpl::addOrUpdateWatch(const std::string& type_url, Watch* watch,
+                                        const std::set<std::string>& resources,
+                                        SubscriptionCallbacks& callbacks,
+                                        std::chrono::milliseconds init_fetch_timeout) {
+  if (watch == nullptr) {
+    return addWatch(type_url, resources, callbacks, init_fetch_timeout);
+  } else {
+    updateWatch(type_url, watch, resources);
+    return watch;
+  }
+}
+
+void NewGrpcMuxImpl::removeWatch(const std::string& type_url, Watch* watch) {
+  updateWatch(type_url, watch, {});
+  auto entry = subscriptions_.find(type_url);
+  if (entry == subscriptions_.end()) {
+    ENVOY_LOG(error, "removeWatch() called for non-existent subscription {}.", type_url);
```

Comment:
No test coverage here, how could this happen? Shouldn't be just an ASSERT?

Reply:
True, it should not happen, and keeping it from never happening only needs the logic within this class to be right. So, I have made this and most other instances in this file into asserts.